hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java

```diff
@@ -0,0 +1,534 @@
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.BucketLayout;
+import org.apache.hadoop.ozone.om.helpers.OmDirectoryInfo;
+import org.apache.hadoop.ozone.om.helpers.WithParentObjectId;
+import org.apache.hadoop.ozone.om.request.file.OMFileRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.TreeMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.NoSuchElementException;
+import java.util.Collection;
+import java.util.Collections;
+
+
+import static org.apache.hadoop.ozone.om.lock.
+    OzoneManagerLock.Resource.BUCKET_LOCK;
+
+/**
+ * Helper class for fetching List Status for a path.
+ */
+public class OzoneListStatusHelper {
+  /**
+   * Interface to get the File Status for a path.
+   */
+  @FunctionalInterface
+  public interface GetFileStatusHelper {
+    OzoneFileStatus apply(OmKeyArgs args, String clientAddress,
+                          boolean skipFileNotFoundError) throws IOException;
+  }
+
+  /**
+   * Interface for iteration of Heap Entries.
+   */
+  public interface ClosableIterator extends Iterator<HeapEntry>, Closeable {
+
+  }
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OzoneListStatusHelper.class);
+
+  private final OMMetadataManager metadataManager;
+  private final long scmBlockSize;
+  private final GetFileStatusHelper getStatusHelper;
+
+  OzoneListStatusHelper(OMMetadataManager metadataManager, long scmBlockSize,
+                        GetFileStatusHelper func) {
+    this.metadataManager = metadataManager;
+    this.scmBlockSize = scmBlockSize;
+    this.getStatusHelper = func;
+  }
+
+  public Collection<OzoneFileStatus> listStatusFSO(OmKeyArgs args,
+      String startKey, long numEntries, String clientAddress)
+      throws IOException {
+    Preconditions.checkNotNull(args, "Key args can not be null");
+
+    boolean listKeysMode = false;
+    final String volumeName = args.getVolumeName();
+    final String bucketName = args.getBucketName();
+    String keyName = args.getKeyName();
+    String prefixKey = keyName;
+
+    String bucketKey = metadataManager.getBucketKey(volumeName, bucketName);
+    OmBucketInfo omBucketInfo =
+        metadataManager.getBucketTable().get(bucketKey);
+    if (omBucketInfo == null) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Volume:{} Bucket:{} does not exist",
+            volumeName, bucketName);
+      }
+      return new ArrayList<>();
+    }
+
+    // Determine if the prefixKey is determined from the startKey
+    // if the keyName is null
+    if (StringUtils.isNotBlank(startKey)) {
+      if (StringUtils.isNotBlank(keyName)) {
+        if (!listKeysMode &&
+            !OzoneFSUtils.arePathsSame(keyName, startKey) &&
+            !OzoneFSUtils.isImmediateChild(keyName, startKey)) {
```

Comment:
StringUtils.startsWith(startKey, keyPrefix) then should proceed. Now it returns saying startKey is not an immediate child.

For ex: listStatus("a1", "a1/b2/d2/d21.tx");

Reply:
ok done.